### DISCUSSION TITLE
Mic not working in composite headset + CDC

### DISCUSSION THREAD

**Discussion by alberto-grl** [hathach/tinyusb#2086] [General]
First time programming of USB so please bear with me.
I'm developing a composite audio and CDC device on an STM32H723. Using STM32Cube for peripherals initializing.

- Mic and speaker using the UAC2 headset example works fine.
- Added the CDC part of the CDC MSC example. Speaker and CDC are OK, mic is somewhat recognized but doesn't work. Strangely, on two different Windows 10 PCs, on one the event log states the peripheral is configured but not started, on the other that "Device Not Migrated due to Partial or Ambiguous Match" but on the next line that the device was started. 
In both cases there is no sound level indicator or level control.

These are the relevant parts of the composite device configuration:

in main.c:
```
while (1)
	{
		tud_task();
		audio_task();
#ifdef COMPOSITE_CDC_AUDIO
		cdc_task();
#endif
	}
```

in usb_descriptors.c:

```
#ifdef COMPOSITE_CDC_AUDIO
#define CONFIG_TOTAL_LEN    	(TUD_CONFIG_DESC_LEN + CFG_TUD_AUDIO * TUD_AUDIO_HEADSET_STEREO_DESC_LEN + TUD_CDC_DESC_LEN)
#else
#define CONFIG_TOTAL_LEN    	(TUD_CONFIG_DESC_LEN + CFG_TUD_AUDIO * TUD_AUDIO_HEADSET_STEREO_DESC_LEN)
#endif

#define EPNUM_AUDIO_IN    0x01
#define EPNUM_AUDIO_OUT   0x01
 #define EPNUM_CDC_NOTIF   0x81
 #define EPNUM_CDC_OUT     0x02
 #define EPNUM_CDC_IN      0x82


// array of pointer to string descriptors
char const* string_desc_arr [] =
{
		(const char[]) { 0x09, 0x04 },  // 0: is supported language is English (0x0409)
		"TinyUSB",                      // 1: Manufacturer
		"TinyUSB headset",              // 2: Product
		"000001",                       // 3: Serials, should use chip ID
		"TinyUSB Speakers",             // 4: Audio Interface
		"TinyUSB Microphone",           // 5: Audio Interface
		"TinyUSB CDC",                 // 6: CDC Interface
};

```
```

uint8_t const desc_configuration[] =
{
		// Config number, interface count, string index, total length, attribute, power in mA
		TUD_CONFIG_DESCRIPTOR(1, ITF_NUM_TOTAL, 0, CONFIG_TOTAL_LEN, 0x00, 100),
#ifdef COMPOSITE_CDC_AUDIO
		 // Interface number, string index, EP notification address and size, EP data address (out, in) and size.
		  TUD_CDC_DESCRIPTOR(4/*ITF_NUM_CDC*/, 4, EPNUM_CDC_NOTIF, 8, EPNUM_CDC_OUT, EPNUM_CDC_IN, 64),
#endif
		// Interface number, string index, EP Out & EP In address, EP size
		TUD_AUDIO_HEADSET_STEREO_DESCRIPTOR(2, EPNUM_AUDIO_OUT, EPNUM_AUDIO_IN | 0x80)
};

```

in usb_descriptors.h:

```
enum
{
  ITF_NUM_AUDIO_CONTROL = 0,
  ITF_NUM_AUDIO_STREAMING_SPK,
  ITF_NUM_AUDIO_STREAMING_MIC,
#ifdef COMPOSITE_CDC_AUDIO
  ITF_NUM_CDC,
  ITF_NUM_CDC_DATA,
#endif
  ITF_NUM_TOTAL
};
```
in tusb_config.h

```
//------------- CLASS -------------//
#define CFG_TUD_AUDIO             1
#ifdef COMPOSITE_CDC_AUDIO
#define CFG_TUD_CDC               1
#else
#define CFG_TUD_CDC               0
#endif
#define CFG_TUD_MSC               0
#define CFG_TUD_HID               0
#define CFG_TUD_MIDI              0
#define CFG_TUD_VENDOR            0

```
Rest of the code is straight from the examples.

Thank you for any hint.

Alberto

**Comment by mkarliner** [comment]
What does your
tud_descriptor_configuration_cb()
look like?

**Comment by alberto-grl** [comment]
Michael,
Thanks for the reply.
I didn't modify it:
```
uint8_t const * tud_descriptor_configuration_cb(uint8_t index)
{
	(void)index; // for multiple configurations
	return desc_configuration;
}
```

Anyway It seems the problem was here:

```
#define EPNUM_AUDIO_IN    0x01
#define EPNUM_AUDIO_OUT   0x01
 #define EPNUM_CDC_NOTIF   0x81
 #define EPNUM_CDC_OUT     0x02
 #define EPNUM_CDC_IN      0x82
```

EP 1 was used by both classes, I tested other EP values but didn't get any result because I didn't change the device descriptors before the test, and Windows kept the previously cached settings. Not exactly a painless experience.
It would be great to have a macro system that auto generates the code above, starting from the number and type of classes required. After all most users just follow my same pattern and painstakingly obtain the same code.


The comment before the TUD_AUDIO_HEADSET_STEREO_DESCRIPTOR call in the headphone example is wrong, parameters are not as stated.
```
// Interface number, string index, EP Out & EP In address, EP size
TUD_AUDIO_HEADSET_STEREO_DESCRIPTOR(2, EPNUM_AUDIO_OUT, EPNUM_AUDIO_IN | 0x80)
```

Now I'm trying to figure how to the mic buffer mechanism works. Not documented, not easy to understand and no examples AFAIK.

Regards,

Alberto

**Reply by mkarliner** [reply]
Yes, the comment is incorrect. I needed two UAC devices and put in the interface number parameter, but there are many other parts of that macro that assume only one device.

I did make a UAC2 + CDC configuration, although it is pretty messy, I'm happy to share it if you need it.

Certainly the documentation is, umm, sparse. However, he tinyusb source code does have some quite extensive comments.

The whole buffer mechanism is confusing, as there is a lot of casting going on.


Mike

**Reply by alberto-grl** [reply]
Mike, thanks for your offer. Right now I have a working example, it performs reasonably well. Still not clear how to use the more advanced kind of buffers, just pushing the exact amount of data to the TX callback for now.

https://github.com/alberto-grl/TinyUSB_UAC_CDC_723

The irony is that a pull request for exactly this was rejected. https://github.com/hathach/tinyusb/pull/1717

Alberto

**Reply by kholia** [reply]
@alberto-grl I have opened https://github.com/hathach/tinyusb/pull/2105 to get a CDC+UAC2 example merged into the TinyUSB repository. Once this example is merged and is a bit stable on RP2040 (via https://github.com/hathach/tinyusb/pull/1802 work), it should provide a decent base to build upon (and then upstream).
